File: year2016/day05/typescript/solution.ts
import { readFileSync } from 'fs';
import { Md5 } from 'ts-md5';

const file: any = readFileSync('../' + (process.argv[2] === 'puzzle' ? 'puzzle' : 'example') + '.txt', 'utf-8');
const input: string = file.toString();
if(process.argv[2] === 'puzzle') {
  console.log(part_1());
  console.log(part_2());
}

export function part_1(): string {
  let [i, pw] = [0, ''];
  while (pw.length < 8) {
    const h = Md5.hashStr(input + i++).toString();
    if (h.startsWith('00000')) pw += h.charAt(5);
  }
  return pw;
}

export function part_2(): string {
  const pw: string[] = Array(8).fill('_');
  for (let i = 0; pw.includes('_'); i++) {
    const hash = Md5.hashStr(input + i).toString();
    if ((hash.startsWith('00000')) && (Number(hash.charAt(5)) >= 0 && Number(hash.charAt(5)) < 8 && pw[Number(hash.charAt(5))] === '_')) pw[Number(hash.charAt(5))] = hash.charAt(6);
  }
  return pw.join('');
}

File: year2016/day05/java/Solution.java
import java.nio.file.*;
import java.util.*;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class Solution {
  private static String PASSWORD;

  public static void main(String[] args) {
    try {
      setup(Files.readString(Paths.get("../puzzle.txt")));
      System.out.println(part1());
      System.out.println(part2());
    } catch(Exception e) {
      e.printStackTrace();
    }
  }

	public static String part1() {
    int i = 0;
    String pw = "";
    while(pw.length() < 8) {
      String h = md5Hash(PASSWORD + i++);
      if (h.startsWith("00000")) {
        pw += h.charAt(5);
      }
    }
    return pw;
  }

  public static String part2() {
    char[] pw = new char[8];
    Arrays.fill(pw, '_');
    for (int i = 0; new String(pw).contains("_"); i++) {
      String hash = md5Hash(PASSWORD + i);
      if (hash.startsWith("00000")) {
        int index = Character.getNumericValue(hash.charAt(5));
        if (index >= 0 && index < 8 && pw[index] == '_') {
          pw[index] = hash.charAt(6);
        }
      }
    }
    return new String(pw);
  }

  public static void setup(String input) {
    PASSWORD = input;
  }

  private static String md5Hash(String input) {
    try {
      MessageDigest md = MessageDigest.getInstance("MD5");
      byte[] messageDigest = md.digest(input.getBytes());
      StringBuilder hexString = new StringBuilder();
      for (byte b : messageDigest) {
        String hex = Integer.toHexString(0xff & b);
        if (hex.length() == 1) {
          hexString.append('0');
        }
        hexString.append(hex);
      }
      return hexString.toString();
    } catch (NoSuchAlgorithmException e) {
      throw new RuntimeException("MD5 algorithm not available", e);
    }
  }
}